src/translator/nth_element.cu

```diff
@@ -373,68 +72,47 @@ public:
     const auto vocabSize = scores->shape()[-1];
     const auto inputN    = scores->shape()[-2];
     const auto dimBatch  = scores->shape()[-4];
+
     ABORT_IF(inputN != (isFirst ? 1 : N), "Input tensor has wrong beam dim??"); // @TODO: Remove isFirst argument altogether
     ABORT_IF(vocabSize > MAX_VOCAB_SIZE, "GetNBestList(): actual vocab size {} exceeds MAX_VOCAB_SIZE of {}", vocabSize, MAX_VOCAB_SIZE);
     ABORT_IF(dimBatch > maxBatchSize_, "GetNBestList(): actual batch size {} exceeds initialization parameter {}", dimBatch, maxBatchSize_);
     ABORT_IF(std::max(N, (size_t)inputN) > maxBeamSize_, "GetNBestList(): actual beam size {} exceeds initialization parameter {}", N, maxBeamSize_);
 
-    const std::vector<size_t> beamSizes(dimBatch, N);
-    std::vector<int> cumulativeBeamSizes(beamSizes.size() + 1, 0);    
-    std::vector<int> batchFirstElementIdxs(beamSizes.size() + 1, 0);
-
-     for(size_t batchIdx = 0; batchIdx < beamSizes.size(); ++batchIdx) {
-#if 1
-      cumulativeBeamSizes[batchIdx + 1] = (batchIdx + 1) * (int)N;
-      batchFirstElementIdxs[batchIdx + 1] += (batchIdx + 1) * inputN * vocabSize;
-      ABORT_IF(cumulativeBeamSizes[batchIdx + 1] != cumulativeBeamSizes[batchIdx] + (int)N, "cumulativeBeamSizes wrong??");
-      ABORT_IF((isFirst ? batchIdx + 1 : cumulativeBeamSizes[batchIdx + 1]) != (batchIdx + 1) * inputN, "inputN wrong??");
-#else
-      cumulativeBeamSizes[batchIdx + 1] = cumulativeBeamSizes[batchIdx] + beamSizes[batchIdx];
-      ABORT_IF(cumulativeBeamSizes[batchIdx + 1] != (batchIdx + 1) * N, "cumulativeBeamSizes wrong??");
-      batchFirstElementIdxs[batchIdx + 1]
-          += ((isFirst) ? (batchIdx + 1) : cumulativeBeamSizes[batchIdx + 1]) * vocabSize;
-      ABORT_IF((isFirst ? batchIdx + 1 : cumulativeBeamSizes[batchIdx + 1]) != (batchIdx + 1) * inputN, "inputN wrong??");
-#endif
-    }
-
     if(scores->type() == Type::float32) {
-      float disabledPathScore = NumericLimits<float>(scores->type()).lowest;
-      selectNBest(scores->data<float>(), batchFirstElementIdxs, cumulativeBeamSizes, disabledPathScore);
+      selectNBest(scores->data<float>(), dimBatch, inputN, N, vocabSize);
+      getPairs<float>(dimBatch * N, outKeys, outCosts);
 #if COMPILE_FP16
     } else if(scores->type() == Type::float16) {
-      float disabledPathScore = NumericLimits<float>(scores->type()).lowest;
-      selectNBest(scores->data<half>(), batchFirstElementIdxs, cumulativeBeamSizes, disabledPathScore);
+      selectNBest(scores->data<half>(), dimBatch, inputN, N, vocabSize);
```

Comment:
What's the difference between `dimBatch` and `beamsPerBatch`?

Reply:
`dimBatch` is the batch dimension and `beamsPerBatch' is the number of hypotheses for each sentence in the batch. I can add a comment in the code.

Reply:
Thanks. This is an incorrect use of the term "beam." [@emjotde -- this is an example of what I mentioned to you last week]. In correct usage, "beam" refers to the set of hypotheses, not a single one. How about rename it to `beamWidths`?

Reply:
I see. I will address this when I address the rest of your feedback. (Thanks for all the helpful feedback btw!)

(Also, leave this open so that I remember to address it. I will close as resolved after.)